contracts/token/ERC1155/ERC1155.sol

```diff
@@ -0,0 +1,234 @@
+        @return The owner's balance of the token type requested
+     */
+    function balanceOf(address owner, uint256 id) public view returns (uint256) {
+        require(owner != address(0), "ERC1155: balance query for the zero address");
+        return _balances[id][owner];
+    }
+
+    /**
+        @dev Get the balance of multiple account/token pairs
+        @param owners The addresses of the token holders
+        @param ids IDs of the tokens
+        @return Balances for each owner and token id pair
+     */
+    function balanceOfBatch(
+        address[] memory owners,
+        uint256[] memory ids
+    )
+        public
+        view
+        returns (uint256[] memory)
+    {
+        require(owners.length == ids.length, "ERC1155: owners and IDs must have same lengths");
+
+        uint256[] memory batchBalances = new uint256[](owners.length);
+
+        for (uint256 i = 0; i < owners.length; ++i) {
+            require(owners[i] != address(0), "ERC1155: some address in batch balance query is zero");
+            batchBalances[i] = _balances[ids[i]][owners[i]];
+        }
+
+        return batchBalances;
+    }
+
+    /**
+     * @dev Sets or unsets the approval of a given operator
+     * An operator is allowed to transfer all tokens of the sender on their behalf
+     * @param operator address to set the approval
+     * @param approved representing the status of the approval to be set
+     */
+    function setApprovalForAll(address operator, bool approved) external {
+        _operatorApprovals[msg.sender][operator] = approved;
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    /**
+        @notice Queries the approval status of an operator for a given owner.
+        @param owner     The owner of the Tokens
+        @param operator  Address of authorized operator
+        @return           True if the operator is approved, false if not
+    */
+    function isApprovedForAll(address owner, address operator) external view returns (bool) {
+        return _operatorApprovals[owner][operator];
+    }
+
+    /**
+        @dev Transfers `value` amount of an `id` from the `from` address to the `to` address specified.
+        Caller must be approved to manage the tokens being transferred out of the `from` account.
+        If `to` is a smart contract, will call `onERC1155Received` on `to` and act appropriately.
+        @param from Source address
+        @param to Target address
+        @param id ID of the token type
+        @param value Transfer amount
+        @param data Data forwarded to `onERC1155Received` if `to` is a contract receiver
+    */
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 value,
+        bytes calldata data
+    )
+        external
+    {
+        require(to != address(0), "ERC1155: target address must be non-zero");
+        require(
+            from == msg.sender || _operatorApprovals[from][msg.sender] == true,
```

Comment:
In line 119 there is a condition which checks if the `msg.sender` has been approved to make the transaction. The functionality assembles to the `isApprovedForAll` function. The problem is that it can't be used because of its visibility (`external`), but it could be made `public`.